src/main.rs

```diff
@@ -1,35 +1,690 @@
+    use super::*;
+
+    #[test]
+    #[should_panic]
+    fn open_empty_box_panic() {
+        let empty_box: GiftBox<()> = GiftBox::Empty;
+        empty_box.open();
+    }
+
+    #[test]
+    fn filling_box() {
+        let filled_box = GiftBox::fill(Some(["Toys", "Candy", "Money"]));
+        assert_eq!(filled_box, GiftBox::Gifts(["Toys", "Candy", "Money"]));
+    }
+
+    #[test]
+    fn open_filled_box() {
+        let filled_box = GiftBox::fill(Some(vec![1, 3, 5, 7]));
+        assert_eq!(filled_box.open(), vec![1, 3, 5, 7]);
+    }
+
+    #[test]
+    fn filling_with_some() {
+        let some_filled_box = GiftBox::fill_keeping_some(Some("Chocolate"));
+        assert_eq!(some_filled_box.open(), Some("Chocolate"));
+    }
+
+    #[test]
+    fn filling_with_none() {
+        let empty_box: GiftBox<()> = GiftBox::fill(None);
+        assert_eq!(empty_box, GiftBox::Empty);
+    }
+
+    #[test]
+    fn wrap_gift_box_with_tag() {
+        let filled_box = GiftBox::fill(Some(["Toys", "Candy", "Money"]));
+        let tag = GiftTag::write(
+            "Bob".to_string(),
+            "Sally".to_string(),
+            "Happy Cake Day!".to_string()
+        );
+        let wrapped_box = filled_box.wrap(
+            Patterns::Polkadots,
+            true,
+            Some(tag)
+        );
+        assert_eq!(
+            wrapped_box,
+            {
+                GiftWrap {
+                    contents: {
+                        GiftBox::Gifts(["Toys", "Candy", "Money"])
+                    },
+                    pattern: Patterns::Polkadots,
+                    has_bow: true,
+                    tag: Some(
+                        GiftTag {
+                            recipient: "Bob".to_string(),
+                            sender: "Sally".to_string(),
+                            message: "Happy Cake Day!".to_string()
+                        }
+                    )
+                }
+            }
+        )
+    }
+
+    #[test]
+    fn wrap_and_unwrap_gift_box() {
+        let filled_box = GiftBox::fill(Some(["Toys", "Candy", "Money"]));
+        let tag = GiftTag::write(
+            "Bob".to_string(),
+            "Sally".to_string(),
+            "Happy Cake Day!".to_string()
+        );
+        let wrapped_box = filled_box.wrap(
+            Patterns::Polkadots,
+            true,
+            Some(tag)
+        );
+        let unwrapped_box = wrapped_box.unwrap();
+        assert_eq!(unwrapped_box, filled_box);
+    }
+
+    #[test]
+    fn wrap_and_open_gift_box_with_tag() {
+        let filled_box = GiftBox::fill(Some(["Toys", "Candy", "Money"]));
+        let tag = GiftTag::write(
+            "Bob".to_string(),
+            "Sally".to_string(),
+            "Happy Cake Day!".to_string()
+        );
+        let wrapped_box = filled_box.wrap(
+            Patterns::Polkadots,
+            true,
+            Some(tag)
+        );
+        assert_eq!(wrapped_box.unwrap().open(), ["Toys", "Candy", "Money"]);
```

Comment:
I would not write this test. The goal of unit tests is to ensure the functionality of as small of units as possible. You have above a test which verifies that GiftBox::open works, and you have a test above that ensures that GiftWrap::unwarp works, so this test is redundant